internal/controller/progressive/progressive.go

```diff
@@ -188,6 +189,12 @@ func ProcessResource(
 		}
 	}
 
+	// determine if we need to replace the Upgrading child with a newer one
+	needsRequeue, err := checkForUpgradeReplacement(ctx, rolloutObject, controller, existingPromotedChild, currentUpgradingChildDef, c)
+	if needsRequeue {
+		return false, common.DefaultRequeueDelay, err
+	}
```

Comment:
added new function `checkForUpgradeReplacement()` to consolidate the logic into a function

Reply:
and now calling it here instead of after we assess failure